pandas-stubs/core/resample.pyi

```diff
@@ -1,56 +1,137 @@
-from typing import Generic
+from typing import (
+    Callable,
+    Generator,
+    Generic,
+    Hashable,
+    Literal,
+    overload,
+)
 
-from pandas import DataFrame
+import numpy as np
+from pandas import (
+    DataFrame,
+    Index,
+    Series,
+)
+from pandas.core.groupby.generic import SeriesGroupBy
 from pandas.core.groupby.groupby import BaseGroupBy
 
-from pandas._typing import NDFrameT
+from pandas._typing import (
+    Axis,
+    NDFrameT,
+    Scalar,
+    npt,
+)
+
+_GroupByFunc = Callable[[Series], Scalar] | Callable[[DataFrame], Series]
+_GroupByFuncTypes = _GroupByFunc | str | list[_GroupByFunc | str]
+_GroupByFuncArgs = _GroupByFuncTypes | dict[Hashable, _GroupByFuncTypes]
+_Interpolation = Literal[
+    "linear",
+    "time",
+    "index",
+    "pad",
+    "nearest",
+    "zero",
+    "slinear",
+    "quadratic",
+    "cubic",
+    "spline",
+    "barycentric",
+    "polynomial",
+    "krogh",
+    "piecewise_polynomial",
+    "spline",
+    "pchip",
+    "akima",
+    "cubicspline",
+    "from_derivatives",
+]
 
 class Resampler(BaseGroupBy, Generic[NDFrameT]):
-    def __init__(self, obj, groupby, axis: int = ..., kind=..., **kwargs) -> None: ...
-    def __getattr__(self, attr: str): ...
-    def __iter__(self): ...
+    def __getattr__(self, attr: str) -> SeriesGroupBy: ...
+    def __iter__(self) -> Generator[tuple[Hashable, NDFrameT], None, None]: ...
     @property
-    def obj(self): ...
+    def obj(self) -> NDFrameT: ...
     @property
-    def ax(self): ...
-    def pipe(self, func, *args, **kwargs): ...
-    def aggregate(self, func=..., *args, **kwargs): ...
+    def ax(self) -> Index: ...
+    def pipe(
+        self,
+        func: Callable | tuple[Callable, str],
+        *args,
+        **kwargs,
+    ) -> Scalar | Series | DataFrame | Resampler: ...
+    def aggregate(
+        self, func: _GroupByFuncArgs | None = ..., *args, **kwargs
+    ) -> Scalar | Series | DataFrame: ...
```

Comment:
Not 100% sure, but with `resample()`, if you are doing `aggregate`, I think the return type will depend upon the function type passed.  See what is done in `SeriesGroupBy` and `DataFrameGroupBy` and see if we can do something similar here.  Same thing may be true for `pipe`

Reply:
It is really hard to get it right since the output depends on the input type and the output type.  I think I'm not going to make any progress on refining this.

Reply:
The two `*GroupBy` have a big advantage since they know the input type. This is generic and so we don't know it here.

Reply:
`Resampler` knows the input type because it is generic.  Not sure why we can't take advantage of that.

Unless the only solution is to do a `SeriesResampler` and `DataFrameResampler` akin to `SeriesGroupBy` and `DataFrameGroupBy`, which I realize would be painful

Reply:
They don't always return the same type.  For example, if using a Dataframe, and Callable[DataFrame, Scalar] you get a Series.  If using a Series, and the same callable, you get a Series.   Could always just ignore that case and say DataFrame always gets DataFrame, and Series always get Series. With that simplification that it could be `NDFramteT`

Reply:
I think (but I am REALLY not sure) you might be able to do something like:
```python
@overload
def aggregate(
        self:Resampler[DataFrame], func: Callable[[DataFrame], Scalar]| None = ..., *args, **kwargs
    ) -> Series: ...
@overload
def aggregate(
        self:Resampler[Series], func: Callable[[DataFrame], Scalar]| None = ..., *args, **kwargs
    ) -> Scalar: ...

Reply:
I think it is too complicated (for me).  Not saying it couldn't be cracked, but it is challenging.  I figured out you can have

DataFrame -> DataFrame | Series, depending on the callable and on the argument type (list or dict can taks a Callable[[DataFrame],Scalar] and get a DataFrame, while just the Callable gets a Series. Similarly you can have Series -> DataFrame | Series, depending on ehther using a string or simple callable (Series), or list of dict (DataFrame).

Reply:
OK.  I guess we'd have to create a bunch of examples that show all the possibilities to see if anything could really be done, or if it is just ends up being that the result is really dynamically typed, then we can't cover it with static typing.